### PR TITLE
Add stackdriver formatter for logrus

### PR DESCRIPTION
**Pull request by babygoat**

### CODE REVIEW

File: logrus.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/genproto/googleapis/devtools/clouderrorreporting/v1beta1"
+	"google.golang.org/genproto/googleapis/logging/type"
+)
+
+type (
+	// Interface for inspecting error objects recursively
+	// Ref: https://github.com/pkg/errors/blob/7f95ac13edff643b8ce5398b6ccab125f8a20c1a/errors.go#L30-L53
```
Comment (nickhsine) [inline logrus.go]: Change ref?
https://godoc.org/github.com/pkg/errors#hdr-Retrieving_the_cause_of_an_error

File: logrus.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	causer interface {
+		Cause() error
+	}
+
+	// Interface for retrieving stack frame for each error object
+	// Ref: https://github.com/pkg/errors/blob/7f95ac13edff643b8ce5398b6ccab125f8a20c1a/errors.go#L68-L73
```
Comment (nickhsine) [inline logrus.go]: Change ref?
https://godoc.org/github.com/pkg/errors#hdr-Retrieving_the_stack_trace_of_an_error_or_wrapper

File: logrus.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+			if fn != nil {
+				file, line := fn.FileLine(pc)
+				lines = append(lines, fmt.Sprintf("%s()\n\t%s:%d +%#x", fn.Name(), file, line, fn.Entry()))
+			}
+		}
+		buf.WriteString(strings.Join(lines, "\n"))
+
+		buffer = buf.Bytes()
+	}
+	return
```
Comment (nickhsine) [inline logrus.go]: What if `err` does not implement `Cause()` and `StackTrace()`?

File: logrus.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func NewStackdriverFormatter(service, version string) *Stackdriver {
```
Comment (nickhsine) [inline logrus.go]: Could we have some comments for exported functions?
Maybe provide some use cases here?

File: logrus.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+		return ltype.LogSeverity_DEBUG
+	default:
+		// Omit intentionally
+	}
+	return ltype.LogSeverity_ERROR
+}
+
+func getGoroutineState() string {
+	stack := make([]byte, 64)
+	stack = stack[:runtime.Stack(stack, false)]
```
Comment (nickhsine) [inline logrus.go]: Could we have some comments here 
to explain what `getGoroutineState` aims to do?

File: stackdriver.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	clouderrorreporting.ErrorEvent
+	Payload map[string]interface{}
+}
+
+func (s *Stackdriver) MarshalJSON() ([]byte, error) {
+	var (
+		m jsonpb.Marshaler
+	)
+
+	unwrap := func(m []byte) []byte {
```
Comment (nickhsine) [inline stackdriver.go]: Could we have some comments here?
Maybe provide some use cases to explain what it aims to do?

**Comment by nickhsine** [comment]
LGTM

File: logrus.go (hunk 3 above)
Comment (babygoat) [reply logrus.go]: Nothing will be printed. This formatter is based on pkg/errors we would need to wrap the information for logging properly.

File: logrus.go (hunk 4 above)
Comment (babygoat) [reply logrus.go]: Please see the example_test.go This is the convention how the go showing examples.
https://blog.golang.org/examples.

File: stackdriver.go (hunk 6 above)
Comment (babygoat) [reply stackdriver.go]: I could add the comments here which might be clear.

**Comment by babygoat** [comment]
Please see example_test.go and let me know if you need any information.

File: logrus.go (hunk 3 above)
Comment (nickhsine) [reply logrus.go]: If nothing will be printed, could we add comments to mention the input `err` should implement `Cause()` and `StackTrace()` in `example_test.go`?

File: logrus.go (hunk 4 above)
Comment (nickhsine) [reply logrus.go]: In example_test.go, 
It's a little bit hard to know what `example` and `test` mean.

**Comment by nickhsine** [comment]
> Please see example_test.go and let me know if you need any information.

Although `example_test.go` shows how to use this pkg briefly, 
the comments I would like to add are the context of what logformatter wants to achieve.
For example,
why we need format the error? -> because we want to show stack in the Stackdriver Error Reporting.

File: logrus.go (hunk 4 above)
Comment (babygoat) [reply logrus.go]: Actually, the ExampleXXX() will provide the example under function XXX() if the godoc is built. This file is indeed tested during `go test` but it always pass as I did not define the output. As I said, this the convention of go. You can see this conventions spread in the standard libraries.

File: logrus.go (hunk 3 above)
Comment (babygoat) [reply logrus.go]: I could add the comment above FormatStack about the usage should be coupled with pkg/errors.

File: logrus.go (hunk 4 above)
Comment (nickhsine) [reply logrus.go]: I know, just want mention that `example` and `test` are not clear for me to read the doc.

File: logrus.go (hunk 1 above)
Comment (babygoat) [reply logrus.go]: done

Comment (babygoat) [reply logrus.go]: done

File: logrus.go (hunk 3 above)
Comment (babygoat) [reply logrus.go]: done.

File: logrus.go (hunk 5 above)
Comment (babygoat) [reply logrus.go]: done

File: stackdriver.go (hunk 6 above)
Comment (babygoat) [reply stackdriver.go]: done

**Comment by babygoat** [comment]
Updated.

**Comment by nickhsine** [comment]
LGTM

**Pull request merged**